~3x speedup, refactoring

philippeitis opened the pull request:
I managed to get the core board solving logic from 0.27s to 0.11 seconds by not adding words to words_info if they're not actual words - these are filtered out later and do not need to be added in the first place.
I also optimized a few other things, such as checking whether a word has a multiplier, and traversing the path.
I also added some pep8 formatting and code organization.

philippeitis committed:
Optimizations, refactoring

Managed to get board solving from 0.27s to 0.11 seconds by not adding words to words_info if they're not actual words - these are filtered out later and do not need to be added in the first place. I also optimized a few other things and simplified the code.

dchen327 commented on ruzzle_bare_minimum.py:
```diff
@@ -49,72 +52,85 @@ def gen_graph():
-def all_combos(board):
+
+    visited[s] = True  # begin DFS, make sure no overlaps
+    path.append(None)
```
What does path.append(None) do?

philippeitis replied:
Instead of appending to the path each time, I append once at the start, set the last element to the specific vertex, and then delete the last element once the graph traversal is complete, as it is either none or not supposed to be in the path.

dchen327 replied:
That's quite smart. I'm going to look over the rest of the code and I'll merge it soon. Are all the double spaces part of PEP8? Like the double space before inline comment, double blank line between functions, etc.

philippeitis replied:
Yes - all of the formatting changes (aside from a bit of spacing for readability inside functions) are PEP8.

philippeitis committed:
Update ruzzle_bare_minimum.py

dchen327 commented on ruzzle_bare_minimum.py:
```diff
@@ -124,31 +140,51 @@ def write_words_to_file(words_info, print_info=False):
+def word_mults_to_int_array():
+    int_word_mults = []
+    for i in range(4):
+        int_word_mults.append([None] * 4)
```
Is this preferred to:
`int_word_mults = [[None] * 4 for _ in range(4)]`

dchen327 commented on ruzzle_bare_minimum.py:
```diff
@@ -124,31 +140,51 @@ def write_words_to_file(words_info, print_info=False):
+    for i in range(4):
+        int_word_mults.append([None] * 4)
+        for j in range(4):
+            int_word_mults[i][j] = int(word_mults[i][j]) if word_mults[i][j] in '23' else 1
```
Would it be better to pass word_mults as a parameter rather than using it globally? Also, would initializing int_word_mults to have 1 instead of None be better, since 1 is the default? This would get rid of the single line if else assignment line.

philippeitis replied:
You would still need to test for values in word_mults, so it doesn't make a significant difference.

philippeitis committed:
Clean up word_mults_to_int_array

dchen327 commented on ruzzle_bare_minimum.py:
```diff
@@ -159,14 +159,15 @@ def get_prefixes():
 
 def word_mults_to_int_array():
-    int_word_mults = []
+    """ Converts word_mults to an array of integers representing the word score multiplies. """
```
There's a small typo: multiplies -> multipliers. Can I edit this commit by myself or do you have to do it? I'm still new to github and stuff.

philippeitis replied:
You can go to the Files Changed tab, go to the file of interest, select the three dots in the upper right corner, and click Edit to edit the file manually.

dchen327 replied:
Thanks for all the help on this project! I'd love to give you some recognition by adding you to a contributors list. What's the easiest way to go about this? Should I just add a contributor section at the bottom of the readme?

philippeitis replied:
Some people do the readme approach, and some people add a CONTRIBUTORS.md file where they list contributors - mostly just a matter of preference.

dchen327 committed:
Update ruzzle_bare_minimum.py

minor edits

philippeitis committed:
Move length check logic

philippeitis committed:
Add uppercase version of TWL06 dictionary

philippeitis committed:
Delete TWL06.txt

philippeitis committed:
Use uppercase dictionary for further 10% time saving.

philippeitis committed:
Add trimmed version of TWL06, add more prefixes

philippeitis committed:
Delete TWL06Upper.txt

philippeitis committed:
Utilize more prefixes, use trimmed file

philippeitis commented:
I've also converted TWL06.txt to TWL06Trimmed.txt - this file contains only words which are <= 12 characters and in their uppercase form - this saves 10% space, and it also allows skipping the .lower() operations.

I've also extended the prefixes to 8 characters, as I have found that this doubles the performance of the dfs:

```
2 to 02: 5.442s
2 to 03: 4.042s
2 to 04: 1.726s
2 to 05: 0.457s
2 to 06: 0.107s
2 to 07: 0.048s
2 to 08: 0.056s
2 to 09: 0.068s
2 to 10: 0.078s
2 to 11: 0.088s
2 to 12: 0.093s
```

(all tests include time to read the prefix files).

The speed up is now 11x for the DFS (0.27s to 0.0236s), and 3x overall, from 0.30s to 0.11s on my machine.

philippeitis committed:
Fix typos

dchen327 commented:
> I've also converted TWL06.txt to TWL06Trimmed.txt - this file contains only words which are <= 12 characters and in their uppercase form - this saves 10% space, and it also allows skipping the .lower() operations.
> 
> I've also extended the prefixes to 8 characters, as I have found that this doubles the performance of the dfs:
> 
> ```
> 2 to 02: 5.442s
> 2 to 03: 4.042s
> 2 to 04: 1.726s
> 2 to 05: 0.457s
> 2 to 06: 0.107s
> 2 to 07: 0.048s
> 2 to 08: 0.056s
> 2 to 09: 0.068s
> 2 to 10: 0.078s
> 2 to 11: 0.088s
> 2 to 12: 0.093s
> ```
> 
> (all tests include time to read the prefix files).
> 
> The speed up is now 11x for the DFS (0.27s to 0.0236s), and 3x overall, from 0.30s to 0.11s on my machine.

My own testing done with timeit.timeit() and 100 iterations varies a lot, but I'm pretty sure the fastest max prefix length to check is 6-8. Your test seems to show a significant difference between 6 and 7 and not a significant difference between 7 and 8, so we can keep 8. How did you run your tests by the way?

philippeitis commented:
> > I've also converted TWL06.txt to TWL06Trimmed.txt - this file contains only words which are <= 12 characters and in their uppercase form - this saves 10% space, and it also allows skipping the .lower() operations.
> > I've also extended the prefixes to 8 characters, as I have found that this doubles the performance of the dfs:
> > ```
> > 2 to 02: 5.442s
> > 2 to 03: 4.042s
> > 2 to 04: 1.726s
> > 2 to 05: 0.457s
> > 2 to 06: 0.107s
> > 2 to 07: 0.048s
> > 2 to 08: 0.056s
> > 2 to 09: 0.068s
> > 2 to 10: 0.078s
> > 2 to 11: 0.088s
> > 2 to 12: 0.093s
> > ```
> > 
> > 
> > (all tests include time to read the prefix files).
> > The speed up is now 11x for the DFS (0.27s to 0.0236s), and 3x overall, from 0.30s to 0.11s on my machine.
> 
> My own testing done with timeit.timeit() and 100 iterations varies a lot, but I'm pretty sure the fastest max prefix length to check is 6-8. Your test seems to show a significant difference between 6 and 7 and not a significant difference between 7 and 8, so we can keep 8. How did you run your tests by the way?

The testing code looked like this. The main source of variance for me was file operations - it's usually a good idea to look at the lowest running time across all runs, as this is the run with the least actual overhead (eg. other processes using the disk). I modified the relevant functions to use LOWER_LIMIT and UPPER_LIMIT.

```
if __name__ == '__main__':
    # configs
    print_info = True  # if True, outputs information about board: number of words and total score
    from timeit import default_timer

    # directory with all necessary files
    main_dir = Path('./')

    # Reading data
    dictionary = get_dict()
    board, word_mults = read_board_from_file()
    word_int_mults = word_mults_to_int_array()
    points = get_points(board, word_mults)

    for i in range(2, 12):

        for j in range(3, 13):
            LOWER_LIMIT = i
            UPPER_LIMIT = j

            start = default_timer()
            prefixes = get_prefixes()

            # Solving
            words_info = []
            all_combos(board, points, word_int_mults)
            words_info = check_words(words_info)
            end = default_timer()
            print(f"{i} to {j}: {end - start} s")
```
